In [72]:
import pandas as pd
import os

In [73]:
RAW_DATA_DIR = os.path.join("..", "data", "raw")

In [ ]:
# loading main csv file
customer_churn_df = pd.read_csv(os.path.join(RAW_DATA_DIR, "telecom_customer_churn.csv"))
# customer_churn_df = pd.read_csv(r"C:\Users\Arpita\Desktop\Customer Churn Prediction\data\raw\telecom_customer_churn.csv")

In [75]:
# loading zipcode_population file
# zipcode_population_df = pd.read_csv(r"C:\Users\Arpita\Desktop\Customer Churn Prediction\data\raw\telecom_zipcode_population.csv")
zipcode_population_df = pd.read_csv(os.path.join(RAW_DATA_DIR,"telecom_zipcode_population.csv"))

In [76]:
# checking information about data types and nulls of the columns in the main file
customer_churn_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 38 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        7043 non-null   str    
 1   Gender                             7043 non-null   str    
 2   Age                                7043 non-null   int64  
 3   Married                            7043 non-null   str    
 4   Number of Dependents               7043 non-null   int64  
 5   City                               7043 non-null   str    
 6   Zip Code                           7043 non-null   int64  
 7   Latitude                           7043 non-null   float64
 8   Longitude                          7043 non-null   float64
 9   Number of Referrals                7043 non-null   int64  
 10  Tenure in Months                   7043 non-null   int64  
 11  Offer                              3166 non-null   str    
 12  Pho

In [77]:
# checking information about data types and nulls of the columns in the zipcode_popultion file
zipcode_population_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1671 entries, 0 to 1670
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   Zip Code    1671 non-null   int64
 1   Population  1671 non-null   int64
dtypes: int64(2)
memory usage: 26.2 KB


In [78]:
# checking top 5 rows of the main file's table
customer_churn_df.head()

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason
0,0002-ORFBO,Female,37,Yes,0,Frazier Park,93225,34.827662,-118.999073,2,9,NaN,Yes,42.39,No,Yes,Cable,16.0,No,Yes,No,Yes,Yes,No,No,Yes,One Year,Yes,Credit Card,65.6,593.30,0.00,0,381.51,974.81,Stayed,NaN,NaN
1,0003-MKNFE,Male,46,No,0,Glendale,91206,34.162515,-118.203869,0,9,NaN,Yes,10.69,Yes,Yes,Cable,10.0,No,No,No,No,No,Yes,Yes,No,Month-to-Month,No,Credit Card,-4.0,542.40,38.33,10,96.21,610.28,Stayed,NaN,NaN
2,0004-TLHLJ,Male,50,No,0,Costa Mesa,92627,33.645672,-117.922613,0,4,Offer E,Yes,33.65,No,Yes,Fiber Optic,30.0,No,No,Yes,No,No,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,73.9,280.85,0.00,0,134.60,415.45,Churned,Competitor,Competitor had better devices
3,0011-IGKFF,Male,78,Yes,0,Martinez,94553,38.014457,-122.115432,1,13,Offer D,Yes,27.82,No,Yes,Fiber Optic,4.0,No,Yes,Yes,No,Yes,Yes,No,Yes,Month-to-Month,Yes,Bank Withdrawal,98.0,1237.85,0.00,0,361.66,1599.51,Churned,Dissatisfaction,Product dissatisfaction
4,0013-EXCHZ,Female,75,Yes,0,Camarillo,93010,34.227846,-119.079903,3,3,NaN,Yes,7.38,No,Yes,Fiber Optic,11.0,No,No,No,Yes,Yes,No,No,Yes,Month-to-Month,Yes,Credit Card,83.9,267.40,0.00,0,22.14,289.54,Churned,Dissatisfaction,Network reliability


In [79]:
# checking the customers who have churned
customer_churn_df[customer_churn_df["Customer Status"] == 'Churned']

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason
2,0004-TLHLJ,Male,50,No,0,Costa Mesa,92627,33.645672,-117.922613,0,4,Offer E,Yes,33.65,No,Yes,Fiber Optic,30.0,No,No,Yes,No,No,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,73.90,280.85,0.0,0,134.60,415.45,Churned,Competitor,Competitor had better devices
3,0011-IGKFF,Male,78,Yes,0,Martinez,94553,38.014457,-122.115432,1,13,Offer D,Yes,27.82,No,Yes,Fiber Optic,4.0,No,Yes,Yes,No,Yes,Yes,No,Yes,Month-to-Month,Yes,Bank Withdrawal,98.00,1237.85,0.0,0,361.66,1599.51,Churned,Dissatisfaction,Product dissatisfaction
4,0013-EXCHZ,Female,75,Yes,0,Camarillo,93010,34.227846,-119.079903,3,3,NaN,Yes,7.38,No,Yes,Fiber Optic,11.0,No,No,No,Yes,Yes,No,No,Yes,Month-to-Month,Yes,Credit Card,83.90,267.40,0.0,0,22.14,289.54,Churned,Dissatisfaction,Network reliability
18,0022-TCJCI,Male,79,No,0,Daly City,94015,37.680844,-122.481310,0,45,NaN,Yes,10.67,No,Yes,DSL,17.0,Yes,No,Yes,No,No,Yes,No,Yes,One Year,No,Credit Card,62.70,2791.50,0.0,0,480.15,3271.65,Churned,Dissatisfaction,Limited range of services
19,0023-HGHWL,Male,67,No,0,Morgan Hill,95037,37.161544,-121.649371,0,1,NaN,No,NaN,NaN,Yes,Cable,9.0,No,No,No,No,No,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,25.10,25.10,0.0,0,0.00,25.10,Churned,Competitor,Competitor made better offer
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7020,9961-JBNMK,Male,79,No,0,Desert Center,92239,33.889605,-115.257009,0,21,NaN,Yes,42.27,No,Yes,Fiber Optic,7.0,No,No,Yes,No,Yes,Yes,No,Yes,Month-to-Month,Yes,Bank Withdrawal,-1.00,2030.30,10.2,0,887.67,2907.77,Churned,Dissatisfaction,Product dissatisfaction
7023,9965-YOKZB,Male,73,No,0,Grizzly Flats,95636,38.636102,-120.522149,0,9,NaN,Yes,10.04,Yes,Yes,Fiber Optic,13.0,No,No,No,No,No,Yes,No,Yes,Month-to-Month,Yes,Bank Withdrawal,83.85,790.15,0.0,0,90.36,880.51,Churned,Dissatisfaction,Network reliability
7036,9985-MWVIX,Female,53,No,0,Hume,93628,36.807595,-118.901544,0,1,Offer E,Yes,42.09,No,Yes,Fiber Optic,9.0,No,No,No,No,No,No,No,Yes,Month-to-Month,Yes,Credit Card,70.15,70.15,0.0,0,42.09,112.24,Churned,Competitor,Competitor had better devices
7037,9986-BONCE,Female,36,No,0,Fallbrook,92028,33.362575,-117.299644,0,4,NaN,Yes,2.01,No,No,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Month-to-Month,No,Bank Withdrawal,20.95,85.50,0.0,0,8.04,93.54,Churned,Competitor,Competitor made better offer


In [80]:
customer_churn_df['Phone Service'].value_counts()

Phone Service
Yes    6361
No      682
Name: count, dtype: int64

In [81]:
# analysing nulls in multiple lines
pd.set_option('display.max_columns', None)
customer_churn_df[customer_churn_df['Multiple Lines'].isnull()]

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason
10,0017-DINOC,Male,47,No,0,Rancho Santa Fe,92091,32.993560,-117.207121,0,54,NaN,No,NaN,NaN,Yes,Cable,10.0,Yes,No,No,Yes,Yes,No,No,Yes,Two Year,No,Credit Card,45.20,2460.55,0.0,0,0.0,2460.55,Stayed,NaN,NaN
14,0019-GFNTW,Female,39,No,0,Los Olivos,93441,34.704340,-120.026090,0,56,NaN,No,NaN,NaN,Yes,DSL,19.0,Yes,Yes,Yes,Yes,No,No,No,Yes,Two Year,No,Bank Withdrawal,45.05,2560.10,0.0,0,0.0,2560.10,Stayed,NaN,NaN
16,0020-JDNXP,Female,52,Yes,1,Point Reyes Station,94956,38.060264,-122.830646,0,34,NaN,No,NaN,NaN,Yes,DSL,20.0,Yes,No,Yes,Yes,Yes,Yes,Yes,Yes,One Year,No,Credit Card,61.25,1993.20,0.0,0,0.0,1993.20,Stayed,NaN,NaN
19,0023-HGHWL,Male,67,No,0,Morgan Hill,95037,37.161544,-121.649371,0,1,NaN,No,NaN,NaN,Yes,Cable,9.0,No,No,No,No,No,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,25.10,25.10,0.0,0,0.0,25.10,Churned,Competitor,Competitor made better offer
25,0032-PGELS,Female,37,Yes,1,Palomar Mountain,92060,33.309852,-116.823091,1,1,NaN,No,NaN,NaN,Yes,DSL,19.0,Yes,No,No,No,No,No,No,Yes,Month-to-Month,No,Bank Withdrawal,30.50,30.50,0.0,0,0.0,30.50,Churned,Attitude,Attitude of service provider
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6979,9906-NHHVC,Female,75,No,0,Macdoel,96058,41.769709,-121.920630,0,65,Offer B,No,NaN,NaN,Yes,Cable,9.0,No,Yes,Yes,No,Yes,Yes,No,Yes,Two Year,Yes,Bank Withdrawal,53.50,3517.90,0.0,0,0.0,3517.90,Stayed,NaN,NaN
6980,9907-SWKKF,Female,69,No,0,Twain,95984,40.022184,-121.062384,0,1,Offer E,No,NaN,NaN,Yes,DSL,21.0,No,No,No,No,No,No,No,Yes,Month-to-Month,No,Credit Card,25.05,25.05,0.0,0,0.0,25.05,Churned,Competitor,Competitor had better devices
6996,9928-BZVLZ,Female,58,No,0,Sunset Beach,90742,33.719221,-118.073596,0,12,Offer D,No,NaN,NaN,Yes,DSL,7.0,Yes,No,Yes,Yes,No,Yes,Yes,Yes,Two Year,No,Credit Card,49.85,552.10,0.0,0,0.0,552.10,Stayed,NaN,NaN
7016,9955-QOPOY,Male,23,Yes,0,Sunnyvale,94089,37.421633,-122.009613,7,69,NaN,No,NaN,NaN,Yes,DSL,42.0,No,Yes,Yes,Yes,Yes,Yes,Yes,Yes,One Year,Yes,Credit Card,60.80,4263.40,0.0,0,0.0,4263.40,Stayed,NaN,NaN


In [82]:
# accoding to dictionary table(reference table), 'Multiple Lines' should be 'No' where 'Phone Service' is 'No'.
# So checking if there is any null in 'Multiple Lines' where 'Phone Service' is 'Yes' and found none.
# Hence, we need to replace these nulls to 'No'
pd.set_option('display.max_columns', None)
customer_churn_df[customer_churn_df['Multiple Lines'].isnull() & customer_churn_df['Phone Service']=='Yes']

C:\Users\Arpita\AppData\Local\Temp\ipykernel_30284\3196305920.py:5: Pandas4Warning: 'and' operations between boolean dtype and str are deprecated and will raise in a future version. Explicitly cast the strings to a boolean dtype before operating instead.
  customer_churn_df[customer_churn_df['Multiple Lines'].isnull() & customer_churn_df['Phone Service']=='Yes']


,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason


In [83]:
# analysing values and nulls of "Internet Type"
customer_churn_df['Internet Type'].value_counts()

Internet Type
Fiber Optic    3035
DSL            1652
Cable           830
Name: count, dtype: int64

In [84]:
# accoding to dictionary table(reference table), 'Internet Type' should be 'None' where 'Internet Service' is 'No'.
# So checking if there is any null in 'Internet Type' where 'Internet Service' is 'Yes' and found none.
# Hence, we need to replace these nulls with 'None'
customer_churn_df[((customer_churn_df['Internet Type'].isnull())) & customer_churn_df['Internet Service']=='Yes']

C:\Users\Arpita\AppData\Local\Temp\ipykernel_30284\1970203847.py:4: Pandas4Warning: 'and' operations between boolean dtype and str are deprecated and will raise in a future version. Explicitly cast the strings to a boolean dtype before operating instead.
  customer_churn_df[((customer_churn_df['Internet Type'].isnull())) & customer_churn_df['Internet Service']=='Yes']


,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason


In [85]:
#analysig nulls of 'Avg Monthly Long Distance Charges'
customer_churn_df[customer_churn_df['Avg Monthly Long Distance Charges'].isnull()]

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason
10,0017-DINOC,Male,47,No,0,Rancho Santa Fe,92091,32.993560,-117.207121,0,54,NaN,No,NaN,NaN,Yes,Cable,10.0,Yes,No,No,Yes,Yes,No,No,Yes,Two Year,No,Credit Card,45.20,2460.55,0.0,0,0.0,2460.55,Stayed,NaN,NaN
14,0019-GFNTW,Female,39,No,0,Los Olivos,93441,34.704340,-120.026090,0,56,NaN,No,NaN,NaN,Yes,DSL,19.0,Yes,Yes,Yes,Yes,No,No,No,Yes,Two Year,No,Bank Withdrawal,45.05,2560.10,0.0,0,0.0,2560.10,Stayed,NaN,NaN
16,0020-JDNXP,Female,52,Yes,1,Point Reyes Station,94956,38.060264,-122.830646,0,34,NaN,No,NaN,NaN,Yes,DSL,20.0,Yes,No,Yes,Yes,Yes,Yes,Yes,Yes,One Year,No,Credit Card,61.25,1993.20,0.0,0,0.0,1993.20,Stayed,NaN,NaN
19,0023-HGHWL,Male,67,No,0,Morgan Hill,95037,37.161544,-121.649371,0,1,NaN,No,NaN,NaN,Yes,Cable,9.0,No,No,No,No,No,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,25.10,25.10,0.0,0,0.0,25.10,Churned,Competitor,Competitor made better offer
25,0032-PGELS,Female,37,Yes,1,Palomar Mountain,92060,33.309852,-116.823091,1,1,NaN,No,NaN,NaN,Yes,DSL,19.0,Yes,No,No,No,No,No,No,Yes,Month-to-Month,No,Bank Withdrawal,30.50,30.50,0.0,0,0.0,30.50,Churned,Attitude,Attitude of service provider
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6979,9906-NHHVC,Female,75,No,0,Macdoel,96058,41.769709,-121.920630,0,65,Offer B,No,NaN,NaN,Yes,Cable,9.0,No,Yes,Yes,No,Yes,Yes,No,Yes,Two Year,Yes,Bank Withdrawal,53.50,3517.90,0.0,0,0.0,3517.90,Stayed,NaN,NaN
6980,9907-SWKKF,Female,69,No,0,Twain,95984,40.022184,-121.062384,0,1,Offer E,No,NaN,NaN,Yes,DSL,21.0,No,No,No,No,No,No,No,Yes,Month-to-Month,No,Credit Card,25.05,25.05,0.0,0,0.0,25.05,Churned,Competitor,Competitor had better devices
6996,9928-BZVLZ,Female,58,No,0,Sunset Beach,90742,33.719221,-118.073596,0,12,Offer D,No,NaN,NaN,Yes,DSL,7.0,Yes,No,Yes,Yes,No,Yes,Yes,Yes,Two Year,No,Credit Card,49.85,552.10,0.0,0,0.0,552.10,Stayed,NaN,NaN
7016,9955-QOPOY,Male,23,Yes,0,Sunnyvale,94089,37.421633,-122.009613,7,69,NaN,No,NaN,NaN,Yes,DSL,42.0,No,Yes,Yes,Yes,Yes,Yes,Yes,Yes,One Year,Yes,Credit Card,60.80,4263.40,0.0,0,0.0,4263.40,Stayed,NaN,NaN


In [86]:
# accoding to dictionary table(reference table), 'Avg Monthly Long Distance Charges' should be '0' where 'Phone Service' is 'No'.
# So checking if there is any null in 'Avg Monthly Long Distance Charges' where 'Phone Service' is 'Yes' and found none.
# Hence, we need to replace these nulls with '0'
customer_churn_df[customer_churn_df['Avg Monthly Long Distance Charges'].isnull() & customer_churn_df['Phone Service']=='Yes']

C:\Users\Arpita\AppData\Local\Temp\ipykernel_30284\1270124287.py:4: Pandas4Warning: 'and' operations between boolean dtype and str are deprecated and will raise in a future version. Explicitly cast the strings to a boolean dtype before operating instead.
  customer_churn_df[customer_churn_df['Avg Monthly Long Distance Charges'].isnull() & customer_churn_df['Phone Service']=='Yes']


,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason


In [87]:
# analysing nulls and values of "Offer"
customer_churn_df["Offer"].value_counts()

Offer
Offer B    824
Offer E    805
Offer D    602
Offer A    520
Offer C    415
Name: count, dtype: int64

Null value analysis findings:
The columns that contains null value are:
- Offer
- Avg Monthly Long Distance Charges
- Multiple Lines
- Internet Type
- Avg Monthly GB Download
- Online Security
- Online Backup
- Device Protection Plan
- Premium Tech Support
- Streaming TV
- Streaming Movies
- Streaming Music
- Unlimited Data
- Churn Category
- Churn Reason

The total rows entries are 7043.

- Now for 'Offer' column, there are 6 acceptable values: None, Offer A, Offer B, Offer C, Offer D and Offer E as per dictionary table. Now in main table, there are these 5 offer values along with nulls. So, we need to replace these nulls with 'None'

- It is given in dictionary table that 'Avg Monthly Long Distance Charges' will be 0 and 'Multiple Lines' will be 'No' if customer is not subscribed to any home phone service, i.e., 'Phone Service' value is 'No'.

So, for 'Phone service' has total 682 'No' value counts.
And after checking data, we found that 'Avg Monthly Long Distance Charges' and 'Multiple Lines' have nulls exactly where 'Phone service' value is 'No'. 
So these total nulls needs to be replaced with either 0 or 'No' as per the column.

- It is given in dictionary table that 'Internet Type' will be 'None', 'Avg Monthly GB Download' will be 0 and 'Online Security', 'Online Backup', 'Device Protection Plan', 'Premium Tech Support', 'Streaming TV', 'Streaming Movies', 'Streaming Music' and 'Unlimited Data' will be 'No'if customer is not subscribed to internet service, i.e., 'Internet Service' value is 'No'.

And after checking data, we found that these columns have nulls exactly where 'Internet service' value is 'No'. 
So these total nulls needs to be replaced with 0, 'No' or 'None' as per the column.

- 'Churn Category' and 'Churn Reason' has nulls where 'Customer Status' is 'Stayed' or 'Joined' which is valid.

In [88]:
customer_churn_df['Offer'] = customer_churn_df['Offer'].fillna('None')
customer_churn_df['Avg Monthly Long Distance Charges'] = customer_churn_df['Avg Monthly Long Distance Charges'].fillna(0)
customer_churn_df['Multiple Lines'] = customer_churn_df['Multiple Lines'].fillna('No')
customer_churn_df['Internet Type'] = customer_churn_df['Internet Type'].fillna('None')
customer_churn_df['Avg Monthly GB Download'] = customer_churn_df['Avg Monthly GB Download'].fillna(0)
customer_churn_df['Online Security'] = customer_churn_df['Online Security'].fillna('No')
customer_churn_df['Online Backup'] = customer_churn_df['Online Backup'].fillna('No')
customer_churn_df['Device Protection Plan'] = customer_churn_df['Device Protection Plan'].fillna('No')
customer_churn_df['Premium Tech Support'] = customer_churn_df['Premium Tech Support'].fillna('No')
customer_churn_df['Streaming Movies'] = customer_churn_df['Streaming Movies'].fillna('No')
customer_churn_df['Streaming Music'] = customer_churn_df['Streaming Music'].fillna('No')
customer_churn_df['Streaming TV'] = customer_churn_df['Streaming TV'].fillna('No')
customer_churn_df['Unlimited Data'] = customer_churn_df['Unlimited Data'].fillna('No')

In [89]:
customer_churn_df['Offer'].isnull().sum()

np.int64(0)

In [90]:
customer_churn_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 38 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        7043 non-null   str    
 1   Gender                             7043 non-null   str    
 2   Age                                7043 non-null   int64  
 3   Married                            7043 non-null   str    
 4   Number of Dependents               7043 non-null   int64  
 5   City                               7043 non-null   str    
 6   Zip Code                           7043 non-null   int64  
 7   Latitude                           7043 non-null   float64
 8   Longitude                          7043 non-null   float64
 9   Number of Referrals                7043 non-null   int64  
 10  Tenure in Months                   7043 non-null   int64  
 11  Offer                              7043 non-null   str    
 12  Pho

In [91]:
customer_churn_df['Gender'].value_counts()

Gender
Male      3555
Female    3488
Name: count, dtype: int64

In [92]:
customer_churn_df['Married'].value_counts()

Married
No     3641
Yes    3402
Name: count, dtype: int64

In [93]:
customer_churn_df['Phone Service'].value_counts()

Phone Service
Yes    6361
No      682
Name: count, dtype: int64

In [94]:
customer_churn_df['Contract'].value_counts()

Contract
Month-to-Month    3610
Two Year          1883
One Year          1550
Name: count, dtype: int64

In [95]:
customer_churn_df['Paperless Billing'].value_counts()

Paperless Billing
Yes    4171
No     2872
Name: count, dtype: int64

In [96]:
customer_churn_df['Payment Method'].value_counts()

Payment Method
Bank Withdrawal    3909
Credit Card        2749
Mailed Check        385
Name: count, dtype: int64

In [97]:
customer_churn_df['Customer Status'].value_counts()

Customer Status
Stayed     4720
Churned    1869
Joined      454
Name: count, dtype: int64

In [98]:
customer_churn_df['Churn Category'].value_counts()

Churn Category
Competitor         841
Dissatisfaction    321
Attitude           314
Price              211
Other              182
Name: count, dtype: int64

In [99]:
customer_churn_df['Churn Reason'].value_counts()

Churn Reason
Competitor had better devices                313
Competitor made better offer                 311
Attitude of support person                   220
Don't know                                   130
Competitor offered more data                 117
Competitor offered higher download speeds    100
Attitude of service provider                  94
Price too high                                78
Product dissatisfaction                       77
Network reliability                           72
Long distance charges                         64
Service dissatisfaction                       63
Moved                                         46
Extra data charges                            39
Limited range of services                     37
Poor expertise of online support              31
Lack of affordable download/upload speed      30
Lack of self-service on Website               29
Poor expertise of phone support               12
Deceased                                       6
Name: c

The biggest reason of churning is competitor, which contributes around 45% churning of customers. According to churned customers, the competitor company is giving better services which particularly includes, they have better devices, better offers, they offers more data and higher download speeds.

In [100]:
customer_churn_df[customer_churn_df.select_dtypes(include='number') < 0].sum()

Customer ID                                       
Gender                                            
Age                                            0.0
Married                                           
Number of Dependents                           0.0
City                                              
Zip Code                                       0.0
Latitude                                       0.0
Longitude                           -843446.323117
Number of Referrals                            0.0
Tenure in Months                               0.0
Offer                                             
Phone Service                                     
Avg Monthly Long Distance Charges              0.0
Multiple Lines                                    
Internet Service                                  
Internet Type                                     
Avg Monthly GB Download                        0.0
Online Security                                   
Online Backup                  

In [101]:
customer_churn_df[customer_churn_df['Monthly Charge']<0]

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason
1,0003-MKNFE,Male,46,No,0,Glendale,91206,34.162515,-118.203869,0,9,None,Yes,10.69,Yes,Yes,Cable,10.0,No,No,No,No,No,Yes,Yes,No,Month-to-Month,No,Credit Card,-4.0,542.40,38.33,10,96.21,610.28,Stayed,NaN,NaN
32,0052-DCKON,Male,28,Yes,0,Los Angeles,90063,34.044271,-118.185237,9,66,None,Yes,43.77,Yes,Yes,Fiber Optic,41.0,Yes,Yes,Yes,Yes,Yes,Yes,Yes,Yes,One Year,No,Bank Withdrawal,-2.0,7942.15,0.00,0,2888.82,10830.97,Stayed,NaN,NaN
170,0274-JKUJR,Male,21,Yes,2,Los Angeles,90040,33.994524,-118.149953,9,7,Offer E,No,0.00,No,Yes,Cable,47.0,Yes,Yes,Yes,No,Yes,Yes,Yes,Yes,Month-to-Month,No,Credit Card,-3.0,465.70,0.00,0,0.00,465.70,Stayed,NaN,NaN
232,0361-HJRDX,Female,38,No,0,Sherman Oaks,91403,34.147149,-118.463365,0,68,Offer A,No,0.00,No,Yes,DSL,28.0,Yes,Yes,Yes,Yes,Yes,Yes,Yes,No,Two Year,No,Credit Card,-8.0,4539.60,0.00,130,0.00,4669.60,Stayed,NaN,NaN
336,0495-ZBNGW,Male,78,Yes,0,Loma Mar,94021,37.266388,-122.263080,1,5,None,Yes,6.66,Yes,Yes,Fiber Optic,14.0,No,No,Yes,No,No,No,No,No,Month-to-Month,Yes,Bank Withdrawal,-1.0,343.95,0.00,20,33.30,397.25,Churned,Competitor,Competitor offered more data
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6894,9793-WECQC,Male,53,No,0,Foresthill,95631,39.031876,-120.811141,0,23,None,Yes,48.21,No,Yes,Fiber Optic,23.0,No,Yes,No,No,Yes,Yes,Yes,Yes,Month-to-Month,No,Credit Card,-1.0,2192.90,0.00,0,1108.83,3301.73,Stayed,NaN,NaN
6918,9821-BESNZ,Male,43,No,0,Northridge,91325,34.236683,-118.517588,0,66,Offer A,Yes,33.02,Yes,Yes,Fiber Optic,6.0,No,Yes,No,Yes,Yes,Yes,Yes,Yes,One Year,Yes,Bank Withdrawal,-7.0,6860.60,44.00,0,2179.32,8995.92,Stayed,NaN,NaN
6976,9898-KZQDZ,Female,65,Yes,0,Lee Vining,93541,37.890145,-119.184087,1,40,None,Yes,49.50,Yes,Yes,Fiber Optic,20.0,No,No,No,No,Yes,Yes,Yes,Yes,Month-to-Month,Yes,Credit Card,-1.0,3640.45,0.00,0,1980.00,5620.45,Churned,Competitor,Competitor had better devices
7020,9961-JBNMK,Male,79,No,0,Desert Center,92239,33.889605,-115.257009,0,21,None,Yes,42.27,No,Yes,Fiber Optic,7.0,No,No,Yes,No,Yes,Yes,No,Yes,Month-to-Month,Yes,Bank Withdrawal,-1.0,2030.30,10.20,0,887.67,2907.77,Churned,Dissatisfaction,Product dissatisfaction


In [102]:
customer_churn_df['Monthly Charge'].describe()

count    7043.000000
mean       63.596131
std        31.204743
min       -10.000000
25%        30.400000
50%        70.050000
75%        89.750000
max       118.750000
Name: Monthly Charge, dtype: float64

In [103]:
neg_df = customer_churn_df[customer_churn_df['Monthly Charge']<0]

In [104]:
neg_df[['Monthly Charge', 'Total Charges', 'Total Refunds', 'Customer Status']]

,Monthly Charge,Total Charges,Total Refunds,Customer Status
1,-4.0,542.40,38.33,Stayed
32,-2.0,7942.15,0.00,Stayed
170,-3.0,465.70,0.00,Stayed
232,-8.0,4539.60,0.00,Stayed
336,-1.0,343.95,0.00,Churned
...,...,...,...,...
6894,-1.0,2192.90,0.00,Stayed
6918,-7.0,6860.60,44.00,Stayed
6976,-1.0,3640.45,0.00,Churned
7020,-1.0,2030.30,10.20,Churned


In [105]:
neg_df["Monthly Charge"].value_counts()

Monthly Charge
-4.0     17
-7.0     16
-3.0     13
-1.0     13
-10.0    13
-8.0     12
-5.0     11
-2.0     10
-9.0      9
-6.0      6
Name: count, dtype: int64

In [106]:
customer_churn_df = customer_churn_df.drop(customer_churn_df[customer_churn_df['Monthly Charge']<0].index)

In [107]:
customer_churn_df.shape

(6923, 38)

Here, I found 120 rows out of 7043 rows where 'Monthly Charge' values were negative which makes no business sense. Also while checking the corresponding 'Total Charges', which were normal, the negative values didn't make any sense. So, I dropped these rows from data. And now the shape of dataframe becomes (6923, 38)

In [108]:
# to check duplicate customer id
# customer_churn_df[customer_churn_df["Customer ID"].value_counts()>1]

In [109]:
ids = customer_churn_df["Customer ID"].value_counts()

In [110]:
ids

Customer ID
0002-ORFBO    1
0004-TLHLJ    1
0011-IGKFF    1
0013-EXCHZ    1
0013-MHZWF    1
             ..
9987-LUTYD    1
9992-RRAMN    1
9992-UJOEL    1
9993-LHIEB    1
9995-HOTOH    1
Name: count, Length: 6923, dtype: int64

In [111]:
duplicate_ids = ids[ids>1].index

In [112]:
customer_churn_df[customer_churn_df['Customer ID'].isin(duplicate_ids)]

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason


In [113]:
#another easy way, direct method
customer_churn_df[customer_churn_df['Customer ID'].duplicated()]

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason


In [114]:
#checking referential integrity of zip codes from 2nd table
customer_churn_df[~(customer_churn_df['Zip Code'].isin(zipcode_population_df['Zip Code']))]

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason


In [ ]:
# creating a new dataset with cleaned data
customer_churn_df.to_csv("customer_churn_df_clean.csv", index = False)

In [116]:
import os
os.getcwd()

'c:\\Users\\Arpita\\Desktop\\Customer Churn Prediction\\notebooks'